File: src/visual-graph/builder/branch-merged-origins.test.ts
import { describe, expect, test } from "vitest";

import { SCOPE_TYPE } from "../../analyzer/scope-type.js";
import { asScopeId } from "../../ir/serialized/scope-id.js";
import type { SerializedIR } from "../../ir/serialized/serialized-ir.js";
import type { SerializedScope } from "../../ir/serialized/serialized-scope.js";
import type { SerializedVariable } from "../../ir/serialized/serialized-variable.js";
import { LANGUAGE } from "../../language.js";
import { AST_TYPE } from "../../parser/ast-type.js";
import { SERIALIZED_IR_VERSION } from "../../serializer/serialized-ir-version.js";
import { branchMergedOrigins } from "./branch-merged-origins.js";
import type { BuilderContext } from "./context.js";
import { baseBlockContext } from "./testing/make-block-context.js";
import { baseScope } from "./testing/make-scope.js";
import { baseWriteOp } from "./testing/make-write-op.js";
import type { WriteOp } from "./write-op.js";

function makeCtx(opts: {
  scopes: readonly SerializedScope[];
  sortedCasesByContainer?: Map<string, readonly SerializedScope[]>;
}): BuilderContext {
  const ir = {
    version: SERIALIZED_IR_VERSION,
    source: { path: "x.ts", language: LANGUAGE.Ts },
    raw: "",
    scopes: opts.scopes,
    variables: [],
    references: [],
    unusedVariableIds: [],
    diagnostics: [],
  } as const satisfies SerializedIR;
  return {
    ir,
    variableMap: new Map<string, SerializedVariable>(),
    scopeMap: new Map(opts.scopes.map((v) => [v.id, v])),
    subgraphOwnerVar: new Map(),
    writeOpsByVariable: new Map(),
    writeOpsByScope: new Map(),
    writeOpByRef: new Map(),
    sortedCasesByContainer: opts.sortedCasesByContainer ?? new Map(),
  };
}

describe("branchMergedOrigins", () => {
  test("returns empty when the branch has no writes", () => {
    const root = { ...baseScope(), id: asScopeId("root") };
    const branch = {
      ...baseScope(),
      id: asScopeId("br"),
      upper: asScopeId("root"),
    };
    const ctx = makeCtx({ scopes: [root, branch] });
    expect(branchMergedOrigins("br", [], ctx)).toEqual([]);
  });

  test("returns the linearly last write when it sits directly in the branch scope", () => {
    const root = { ...baseScope(), id: asScopeId("root") };
    const branch = {
      ...baseScope(),
      id: asScopeId("br"),
      upper: asScopeId("root"),
    };
    const op = { ...baseWriteOp(), refId: "rA", offset: 10, scopeId: "br" };
    const ctx = makeCtx({ scopes: [root, branch] });
    expect(branchMergedOrigins("br", [op], ctx)).toEqual(["wr_rA"]);
  });

  test("recurses into nested if/else under the outer branch and collects both arms", () => {
    const root = { ...baseScope(), id: asScopeId("root") };
    const outer = {
      ...baseScope(),
      id: asScopeId("outer"),
      upper: asScopeId("root"),
    };
    const cons = {
      ...baseScope(),
      id: asScopeId("cons"),
      upper: asScopeId("outer"),
      blockContext: { ...baseBlockContext(), parentSpanOffset: 50 },
    };
    const alt = {
      ...baseScope(),
      id: asScopeId("alt"),
      upper: asScopeId("outer"),
      blockContext: {
        ...baseBlockContext(),
        key: "alternate",
        parentSpanOffset: 50,
      },
    };
    const opCons: WriteOp = {
      ...baseWriteOp(),
      refId: "rCons",
      offset: 60,
      scopeId: "cons",
    };
    const opAlt: WriteOp = {
      ...baseWriteOp(),
      refId: "rAlt",
      offset: 70,
      scopeId: "alt",
    };
    const ctx = makeCtx({ scopes: [root, outer, cons, alt] });
    expect(new Set(branchMergedOrigins("outer", [opCons, opAlt], ctx))).toEqual(
      new Set(["wr_rCons", "wr_rAlt"]),
    );
  });

  test("recurses into a nested switch and collects all reachable cases", () => {
    const root = { ...baseScope(), id: asScopeId("root") };
    const outer = {
      ...baseScope(),
      id: asScopeId("outer"),
      upper: asScopeId("root"),
    };
    const switchScope = {
      ...baseScope(),
      id: asScopeId("switch"),
      type: SCOPE_TYPE.Switch,
      upper: asScopeId("outer"),
    };
    const c1 = {
      ...baseScope(),
      id: asScopeId("c1"),
      upper: asScopeId("switch"),
      blockContext: {
        ...baseBlockContext(),
        parentType: AST_TYPE.SwitchStatement,
        key: "cases",
        parentSpanOffset: 100,
      },
    };
    const c2 = {
      ...baseScope(),
      id: asScopeId("c2"),
      upper: asScopeId("switch"),
      blockContext: {
        ...baseBlockContext(),
        parentType: AST_TYPE.SwitchStatement,
        key: "cases",
        parentSpanOffset: 100,
      },
    };
    const opC1 = { ...baseWriteOp(), refId: "rC1", offset: 110, scopeId: "c1" };
    const opC2 = { ...baseWriteOp(), refId: "rC2", offset: 120, scopeId: "c2" };
    const containerKey = "switch:switch:100";
    const ctx = makeCtx({
      scopes: [root, outer, switchScope, c1, c2],
      sortedCasesByContainer: new Map([[containerKey, [c1, c2]]]),
    });
    expect(new Set(branchMergedOrigins("outer", [opC1, opC2], ctx))).toEqual(
      new Set(["wr_rC1", "wr_rC2"]),
    );
  });

  test("excludes a switch case that exits the function", () => {
    const root = { ...baseScope(), id: asScopeId("root") };
    const outer = {
      ...baseScope(),
      id: asScopeId("outer"),
      upper: asScopeId("root"),
    };
    const switchScope = {
      ...baseScope(),
      id: asScopeId("switch"),
      type: SCOPE_TYPE.Switch,
      upper: asScopeId("outer"),
    };
    const c1 = {
      ...baseScope(),
      id: asScopeId("c1"),
      upper: asScopeId("switch"),
      blockContext: {
        ...baseBlockContext(),
        parentType: AST_TYPE.SwitchStatement,
        key: "cases",
        parentSpanOffset: 100,
      },
      exitsFunction: true,
    };
    const c2 = {
      ...baseScope(),
      id: asScopeId("c2"),
      upper: asScopeId("switch"),
      blockContext: {
        ...baseBlockContext(),
        parentType: AST_TYPE.SwitchStatement,
        key: "cases",
        parentSpanOffset: 100,
      },
    };
    const opC1 = { ...baseWriteOp(), refId: "rC1", offset: 110, scopeId: "c1" };
    const opC2 = { ...baseWriteOp(), refId: "rC2", offset: 120, scopeId: "c2" };
    const containerKey = "switch:switch:100";
    const ctx = makeCtx({
      scopes: [root, outer, switchScope, c1, c2],
      sortedCasesByContainer: new Map([[containerKey, [c1, c2]]]),
    });
    expect(branchMergedOrigins("outer", [opC1, opC2], ctx)).toEqual(["wr_rC2"]);
  });

  test("descends through 3-level nesting (writes in middle.cons and leaf.cons/alt)", () => {
    // outer
    // ├── midCons writes opMid                  (consequent of outer-of-outer if)
    // └── midAlt
    //     └── inner if/else
    //         ├── leafCons writes opLeafC
    //         └── leafAlt  writes opLeafA
    //
    // After outer's case runs, the reachable last writers are:
    // - midCons path: opMid
    // - midAlt path: leafCons OR leafAlt (inner if has alternate, so always
    //   one of them runs; midAlt itself never writes the variable directly)
    // -> merged origins: {opMid, opLeafC, opLeafA}.
    const root = { ...baseScope(), id: asScopeId("root") };
    const outer = {
      ...baseScope(),
      id: asScopeId("outer"),
      upper: asScopeId("root"),
    };
    const middleCons = {
      ...baseScope(),
      id: asScopeId("midCons"),
      upper: asScopeId("outer"),
      blockContext: { ...baseBlockContext(), parentSpanOffset: 50 },
    };
    const middleAlt = {
      ...baseScope(),
      id: asScopeId("midAlt"),
      upper: asScopeId("outer"),
      blockContext: {
        ...baseBlockContext(),
        key: "alternate",
        parentSpanOffset: 50,
      },
    };
    const leafCons = {
      ...baseScope(),
      id: asScopeId("leafCons"),
      upper: asScopeId("midAlt"),
      blockContext: { ...baseBlockContext(), parentSpanOffset: 70 },
    };
    const leafAlt = {
      ...baseScope(),
      id: asScopeId("leafAlt"),
      upper: asScopeId("midAlt"),
      blockContext: {
        ...baseBlockContext(),
        key: "alternate",
        parentSpanOffset: 70,
      },
    };
    const opMidCons = {
      ...baseWriteOp(),
      refId: "rMid",
      offset: 60,
      scopeId: "midCons",
    };
    const opLeafCons = {
      ...baseWriteOp(),
      refId: "rLeafC",
      offset: 80,
      scopeId: "leafCons",
    };
    const opLeafAlt = {
      ...baseWriteOp(),
      refId: "rLeafA",
      offset: 90,
      scopeId: "leafAlt",
    };
    const ctx = makeCtx({
      scopes: [root, outer, middleCons, middleAlt, leafCons, leafAlt],
    });
    expect(
      new Set(
        branchMergedOrigins("outer", [opMidCons, opLeafCons, opLeafAlt], ctx),
      ),
    ).toEqual(new Set(["wr_rMid", "wr_rLeafC", "wr_rLeafA"]));
  });

  test("inner if without alternate keeps the pre-inner write reachable", () => {
    // outer (case-like) writes opPre directly, then runs an `if (cond)`
    // that has no else clause.
    //   case 0:
    //     x = pre        // opPre, scopeId = outer
    //     if (cond) {
    //       x = inner    // opInner, scopeId = innerCons
    //     }
    //     break;
    // Without the fallback, only opInner would be returned, but since the
    // inner-if can be skipped at runtime, opPre is still a possible last
    // writer and must remain in the merged origins.
    const root = { ...baseScope(), id: asScopeId("root") };
    const outer = {
      ...baseScope(),
      id: asScopeId("outer"),
      upper: asScopeId("root"),
    };
    const innerCons = {
      ...baseScope(),
      id: asScopeId("innerCons"),
      upper: asScopeId("outer"),
      blockContext: { ...baseBlockContext(), parentSpanOffset: 100 },
    };
    const opPre = {
      ...baseWriteOp(),
      refId: "rPre",
      offset: 60,
      scopeId: "outer",
    };
    const opInner = {
      ...baseWriteOp(),
      refId: "rInner",
      offset: 110,
      scopeId: "innerCons",
    };
    const ctx = makeCtx({ scopes: [root, outer, innerCons] });
    expect(
      new Set(branchMergedOrigins("outer", [opPre, opInner], ctx)),
    ).toEqual(new Set(["wr_rPre", "wr_rInner"]));
  });
});

File: crates/unsnarl-visual-graph/src/builder/branch_merged_origins_test.rs
//! Sibling tests for [`branch_merged_origins`].

use std::collections::HashSet;

use unsnarl_ir::scope_type::ScopeType;
use unsnarl_ir::serialized::SerializedScope;
use unsnarl_oxc_parity::AstType;

use super::branch_merged_origins;
use crate::builder::testing::{
    base_builder_context, base_serialized_scope, base_write_op, case_clause_block_context,
    empty_serialized_ir, other_block_context, scope_id,
};
use crate::builder::write_op::WriteOp;

fn write_op_at(ref_id: &str, offset: u32, scope_id_str: &str) -> WriteOp {
    WriteOp {
        ref_id: ref_id.to_string(),
        offset,
        scope_id: scope_id_str.to_string(),
        ..base_write_op()
    }
}

fn scope_with_upper(id: &str, upper: &str) -> SerializedScope {
    let mut s = base_serialized_scope(id);
    s.upper = Some(scope_id(upper));
    s
}

fn scope_with_block_context(
    id: &str,
    upper: &str,
    parent_type: AstType,
    key: &str,
    parent_span_offset: u32,
) -> SerializedScope {
    let mut s = scope_with_upper(id, upper);
    s.block_context = Some(other_block_context(
        parent_type,
        key,
        parent_span_offset,
        None,
    ));
    s
}

fn case_scope(id: &str, upper: &str, parent_span_offset: u32) -> SerializedScope {
    let mut s = scope_with_upper(id, upper);
    s.block_context = Some(case_clause_block_context(
        AstType::SwitchStatement,
        "cases",
        parent_span_offset,
        None,
    ));
    s
}

#[test]
fn returns_empty_when_the_branch_has_no_writes() {
    let mut ir = empty_serialized_ir();
    ir.scopes.push(base_serialized_scope("root"));
    ir.scopes.push(scope_with_upper("br", "root"));
    let ctx = base_builder_context(&ir);
    assert!(branch_merged_origins("br", &[], &ctx).is_empty());
}

#[test]
fn returns_linearly_last_write_when_it_sits_directly_in_branch_scope() {
    let mut ir = empty_serialized_ir();
    ir.scopes.push(base_serialized_scope("root"));
    ir.scopes.push(scope_with_upper("br", "root"));
    let ctx = base_builder_context(&ir);
    let op = write_op_at("rA", 10, "br");
    assert_eq!(branch_merged_origins("br", &[op], &ctx), vec!["wr_rA"]);
}

#[test]
fn recurses_into_nested_if_else_and_collects_both_arms() {
    let mut ir = empty_serialized_ir();
    ir.scopes.push(base_serialized_scope("root"));
    ir.scopes.push(scope_with_upper("outer", "root"));
    ir.scopes.push(scope_with_block_context(
        "cons",
        "outer",
        AstType::IfStatement,
        "consequent",
        50,
    ));
    ir.scopes.push(scope_with_block_context(
        "alt",
        "outer",
        AstType::IfStatement,
        "alternate",
        50,
    ));
    let ctx = base_builder_context(&ir);
    let op_cons = write_op_at("rCons", 60, "cons");
    let op_alt = write_op_at("rAlt", 70, "alt");
    let got: HashSet<String> = branch_merged_origins("outer", &[op_cons, op_alt], &ctx)
        .into_iter()
        .collect();
    let expected: HashSet<String> = ["wr_rCons", "wr_rAlt"]
        .iter()
        .map(|s| s.to_string())
        .collect();
    assert_eq!(got, expected);
}

#[test]
fn recurses_into_nested_switch_and_collects_all_reachable_cases() {
    let mut ir = empty_serialized_ir();
    ir.scopes.push(base_serialized_scope("root"));
    ir.scopes.push(scope_with_upper("outer", "root"));
    let mut switch_scope = scope_with_upper("switch", "outer");
    switch_scope.r#type = ScopeType::Switch;
    ir.scopes.push(switch_scope);
    ir.scopes.push(case_scope("c1", "switch", 100));
    ir.scopes.push(case_scope("c2", "switch", 100));
    let mut ctx = base_builder_context(&ir);
    let container_key = "switch:switch:100".to_string();
    let cases: Vec<&SerializedScope> = ir
        .scopes
        .iter()
        .filter(|s| s.id.value() == "c1" || s.id.value() == "c2")
        .collect();
    ctx.sorted_cases_by_container.insert(container_key, cases);
    let op_c1 = write_op_at("rC1", 110, "c1");
    let op_c2 = write_op_at("rC2", 120, "c2");
    let got: HashSet<String> = branch_merged_origins("outer", &[op_c1, op_c2], &ctx)
        .into_iter()
        .collect();
    let expected: HashSet<String> = ["wr_rC1", "wr_rC2"].iter().map(|s| s.to_string()).collect();
    assert_eq!(got, expected);
}

#[test]
fn excludes_switch_case_that_exits_function() {
    let mut ir = empty_serialized_ir();
    ir.scopes.push(base_serialized_scope("root"));
    ir.scopes.push(scope_with_upper("outer", "root"));
    let mut switch_scope = scope_with_upper("switch", "outer");
    switch_scope.r#type = ScopeType::Switch;
    ir.scopes.push(switch_scope);
    let mut c1 = case_scope("c1", "switch", 100);
    c1.exits_function = true;
    ir.scopes.push(c1);
    ir.scopes.push(case_scope("c2", "switch", 100));
    let mut ctx = base_builder_context(&ir);
    let container_key = "switch:switch:100".to_string();
    let cases: Vec<&SerializedScope> = ir
        .scopes
        .iter()
        .filter(|s| s.id.value() == "c1" || s.id.value() == "c2")
        .collect();
    ctx.sorted_cases_by_container.insert(container_key, cases);
    let op_c1 = write_op_at("rC1", 110, "c1");
    let op_c2 = write_op_at("rC2", 120, "c2");
    assert_eq!(
        branch_merged_origins("outer", &[op_c1, op_c2], &ctx),
        vec!["wr_rC2"]
    );
}

#[test]
fn descends_through_three_level_nesting() {
    let mut ir = empty_serialized_ir();
    ir.scopes.push(base_serialized_scope("root"));
    ir.scopes.push(scope_with_upper("outer", "root"));
    ir.scopes.push(scope_with_block_context(
        "midCons",
        "outer",
        AstType::IfStatement,
        "consequent",
        50,
    ));
    ir.scopes.push(scope_with_block_context(
        "midAlt",
        "outer",
        AstType::IfStatement,
        "alternate",
        50,
    ));
    ir.scopes.push(scope_with_block_context(
        "leafCons",
        "midAlt",
        AstType::IfStatement,
        "consequent",
        70,
    ));
    ir.scopes.push(scope_with_block_context(
        "leafAlt",
        "midAlt",
        AstType::IfStatement,
        "alternate",
        70,
    ));
    let ctx = base_builder_context(&ir);
    let op_mid = write_op_at("rMid", 60, "midCons");
    let op_leaf_c = write_op_at("rLeafC", 80, "leafCons");
    let op_leaf_a = write_op_at("rLeafA", 90, "leafAlt");
    let got: HashSet<String> =
        branch_merged_origins("outer", &[op_mid, op_leaf_c, op_leaf_a], &ctx)
            .into_iter()
            .collect();
    let expected: HashSet<String> = ["wr_rMid", "wr_rLeafC", "wr_rLeafA"]
        .iter()
        .map(|s| s.to_string())
        .collect();
    assert_eq!(got, expected);
}

#[test]
fn inner_if_without_alternate_keeps_pre_inner_write_reachable() {
    let mut ir = empty_serialized_ir();
    ir.scopes.push(base_serialized_scope("root"));
    ir.scopes.push(scope_with_upper("outer", "root"));
    ir.scopes.push(scope_with_block_context(
        "innerCons",
        "outer",
        AstType::IfStatement,
        "consequent",
        100,
    ));
    let ctx = base_builder_context(&ir);
    let op_pre = write_op_at("rPre", 60, "outer");
    let op_inner = write_op_at("rInner", 110, "innerCons");
    let got: HashSet<String> = branch_merged_origins("outer", &[op_pre, op_inner], &ctx)
        .into_iter()
        .collect();
    let expected: HashSet<String> = ["wr_rPre", "wr_rInner"]
        .iter()
        .map(|s| s.to_string())
        .collect();
    assert_eq!(got, expected);
}
